### ISSUE TITLE
More performant codegen for logical expressions

### ISSUE THREAD

**Issue by MaxGraey** [AssemblyScript/assemblyscript#1149]
It seems logical expressions much better compile use blocks (1) like:
```ts
export function f3(x: number): number {
  do {
    if (x === 1) {
      break;
    }
    if (x === 2) {
      break;
    }
    return f3(x - 1) + f3(x - 2);
  } while (false);
  return 1;
}
```
instead (2):
```ts
export function f1(x: number): number {
  if (x === 1 || x === 2) {
    return 1;
  }
  return f1(x - 1) + f1(x - 2);
}
```

Benchmark: https://webassembly.studio/?f=n55wmjm5b3

### Chrome 80:
| fib ver   |   time  |
|---------|----------------------------------|
| _f64.f1_ |   _150ms (default with f64 ver)_ (2)
| f64.f2  |  139ms
| **f64.f3** | **124ms (wrapped by block i32 ver)** (1)
| f64.f4 |    134ms
| _i32.i1_ |   _121ms (default with i32 ver)_ (2)
| **i32.i3** |  **112ms (wrapped by block i32 ver)** (1)
| _f64.js_ |  _190ms (javascript)_

### Firefox 73
| fib ver   |   time  |
|---------|----------------------------------|
| _f64.f1_ |   _133ms (default with f64 ver)_ (2)
| f64.f2  |  143ms
| **f64.f3** | **127ms (wrapped by block i32 ver)** (1)
| f64.f4 |    128ms
| _i32.i1_ |   _110ms (default with i32 ver)_ (2)
| **i32.i3** |  **101ms (wrapped by block i32 ver)** (1)
| _f64.js_ |  _117ms (javascript)_

Codegen for (1):
```wat
(func $f3 (export "f3") (type $t0) (param $p0 f64) (result f64)
    block $B0
      get_local $p0
      f64.const 0x1p+0 (;=1;)
      f64.eq
      br_if $B0
      get_local $p0
      f64.const 0x1p+1 (;=2;)
      f64.eq
      br_if $B0
      get_local $p0
      f64.const 0x1p+0 (;=1;)
      f64.sub
      call $f3
      get_local $p0
      f64.const 0x1p+1 (;=2;)
      f64.sub
      call $f3
      f64.add
      return
    end
    f64.const 0x1p+0 (;=1;))
```

Codegen for (2):
```wat
(func $f1 (export "f1") (type $t0) (param $p0 f64) (result f64)
    i32.const 1
    get_local $p0
    f64.const 0x1p+1 (;=2;)
    f64.eq
    get_local $p0
    f64.const 0x1p+0 (;=1;)
    f64.eq
    select
    if $I0
      f64.const 0x1p+0 (;=1;)
      return
    end
    get_local $p0
    f64.const 0x1p+0 (;=1;)
    f64.sub
    call $f1
    get_local $p0
    f64.const 0x1p+1 (;=2;)
    f64.sub
    call $f1
    f64.add)
```

**Comment by jtenner** [comment]
Do you guys need some help identifying places where we should change the std lib?

**Comment by jtenner** [comment]
also, I'm curious if this is faster? 
```ts
if (i32(x == 1) | i32(x == 2)) break;
```

**Comment by MaxGraey** [comment]
No, I guess it should be done by compiler. But need more examples. It seems LLVM also prefer block-scoped variant

**Comment by dcodeIO** [comment]
Current implementation tries to do both the bool (`if (a && b)`) and the value case (`var c = a && b`) as one, with the latter complicating matters a bit. Somewhat related to https://github.com/AssemblyScript/assemblyscript/issues/1136, in that the current value rules in logical expressions are weird anyway and need a rework.

**Comment by dcodeIO** [comment]
@MaxGraey Do you think that this can/should be solved on the Binaryen side? If not, there have been some changes to codegen here lately when exclusively interested in boolean values. Do you know if these did help?

**Comment by MaxGraey** [comment]
I don't know yet. Plz leave this open

**Comment by dcodeIO** [comment]
Sure, no problem :)
